AK/MemoryStream.h

```diff
@@ -252,7 +252,7 @@ class DuplexMemoryStream final : public DuplexStream {
         size_t nread = 0;
         while (bytes.size() - nread > 0 && m_write_offset - m_read_offset - nread > 0) {
             const auto chunk_index = (m_read_offset - m_base_offset + nread) / chunk_size;
-            const auto chunk_bytes = m_chunks[chunk_index].bytes().slice(m_read_offset % chunk_size).trim(m_write_offset - m_read_offset - nread);
+            const auto chunk_bytes = m_chunks[chunk_index].bytes().slice((m_read_offset + nread) % chunk_size).trim(m_write_offset - m_read_offset - nread);
```

Comment:
Also cc @asynts 
Test case is...not minimal: https://github.com/alimpfard/aoc2020-shell/blob/1036d3833b601401ae4848ffbcbb59700dd21346/4.sh#L76

Reply:
I have trouble wrapping my head around all of these offset calculations but I think you are right. If `m_read_offset % chunk_size` is not zero and a second chunk is read, it would be sliced at the wrong offset.

This should be a test case, but for some reason `stream << 1 << input << 1` hangs, seems like it's totally broken. I'll debug that tomorrow.

~~~c++
TEST_CASE(offset_calculation_error_regression)
{
    [[maybe_unused]] Array<u8, DuplexMemoryStream::chunk_size> input, output;
    input.span().fill(0);

    DuplexMemoryStream stream;
    stream << 1 << input << 1;

    stream.discard_or_error(4);
    stream.read(output);

    EXPECT(compare(input, output));
}
~~~

Reply:
(This was resolved in #4371.)

Reply:
Oh yeah, silly me :)